rfd/0112-automatic-database-users.md

```diff
@@ -0,0 +1,317 @@
+provision database users separately when onboarding new users.
+
+Adding support for on-demand database user provisioning, similar to [automatic
+Linux user provisioning](https://github.com/gravitational/teleport/blob/master/rfd/0057-automatic-user-provisioning.md),
+will allow database access users to not worry about preconfiguring each
+individual user within the database and will simplify user on-/offboarding and
+give ability to control permissions via IdP user traits.
+
+## Scope
+
+The RFD will cover self-hosted and RDS versions of PostgreSQL but the same
+approach can then be extended to other protocols.
+
+## High-level flow
+
+Let's first define the high-level flow of how on-demand user provisioning
+will work:
+
+- Users will designate an existing user in their database as "adminUser"
+  which will have permissions to create/delete other users and assign them
+  privileges.
+- When receiving a connection for the user that should be auto-created,
+  Teleport database service will first connect to the database as "adminUser"
+  and create it.
+- Once the user completes the session, Teleport database service will connect
+  as "adminUser" again and disable it if it doesn't have any other active
+  sessions.
+
+Now let's dive into details.
+
+## Details
+
+### Configuration changes
+
+To support this functionality, Teleport database service needs to have access
+to the database as a user that has permissions to create users and assign them
+roles.
+
+To avoid asking users to provide static credentials for the admin user,
+Teleport will use the same authentication mechanism for it as for any other
+user connecting to the database (i.e. X.509 for self-hosted, IAM for RDS, etc).
+This means that the "adminUser" will still need to be configured first to
+allow Teleport to use it. The permissions that "adminUser" will need to hold
+within the database will be defined below.
+
+Database resource spec for both static config and dynamic resources will be
+updated to allow users to specify which "adminUser" Teleport should use for
+auto-provisioning:
+
+```yaml
+kind: "db"
+version: "v3"
+metadata:
+  name: "example"
+spec:
+  protocol: "postgres"
+  uri: "localhost:5432"
+  # The adminUser section is intentionally nested for future extensibility.
+  adminUser:
+    name: "postgres"
+```
+
+### Role changes
+
+The role resource will be updated to include additional options/fields:
+
+- `options.create_db_user` will indicate whether allowed `db_users` from this
+  role should be auto-created, similar to Linux user provisioning.
+- `allow.db_roles` will contain database role names the created users should
+  be granted within the database. Will support IdP trait templating.
+
+```yaml
+kind: "role"
+version: "v5"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user: true
+  allow:
+    db_users: ["alice", "bob", "{{external.db_users}}"]
+    db_roles: ["reader", "writer", "{{external.db_roles}}"]
```

Comment:
I can imagine that `db_roles` can be defined in defined roles. Do we plan support user_roles.db_roles Union ?

Reply:
My plan was to take `db_roles` from your roles that match the database you're connecting to and have `create_db_user` setting. If you have multiple such roles in your roleset, then yes, a union.